recursive-methods.rb

```diff
@@ -1,36 +1,258 @@
+  end
+end
+
+n = 4
+if bunny(4) == 8
+  puts "Bunny alg: Right!"
+else
+  puts "Bunny str alg: BUG - #{bunny(n)}"
+end
+
+#time comp: O(n), where n is the number of bunnies; computational time grows at same rate.
+#space comp: O(n), n is number of bunnies; required space in memory grows at same rate as number of bunnnies.
+
+# Write #5
+# nested(s) Write a method nested that accepts a string of only parenthesis and then returns if those parenthesis are properly nested. You may assume that no non-paren characters will be passed to this method.
+# e.g. nested("((()))") = true
+# e.g. nested("())") = false Challenge: Try doing this without creating new strings in the process of solving the problem.
+
+def nested(s, i=0, j=(-1))
+  base_case = s.length / 2
+  #ensure even length of characters
+  if s.length % 2 != 0
+    return false
+  else
+    if s.length == 0 || i == base_case
+      return true
+    else
+      if (s[i] == "(") && (s[j] == ")")
+        return nested(s, i+1, j-1)
+      else
+        return false
+      end
+    end
+  end
+end
+
+#time comp: where n is the number of characters in the string, it takes O(n/2) time to compute, so O(n)
+#space comp: where n is the number of characters in the string, it takes O(n/2) stack calls, so O(n)
+
+s = "((()))"
+if nested(s) == true
+  puts "Nested alg: Right!"
+else
+  puts "Nested alg: BUG - #{nested(s)}"
 end
 
-def nested(s)
-  puts "Not implemented."
+s = "((())"
+if nested(s) == false
+  puts "Nested alg 2: Right!"
+else
+  puts "Nested alg alg 2: BUG - #{nested(s)}"
 end
 
-def search(array, value)
-  puts "Not implemented."
+
+# Write #6
+# search(array, value) Write a method search that accepts an unsorted array of integers and an integer value to find and then returns true if the value if found in the unsorted array and false otherwise. Make the algorithm recursive.
+# e.g. search([34, 45, 21, 4, 67], 4) should return true
+# e.g. search([24, 51, 11], 4) should return false
+# e.g. search([], 4) should return false
+
+def search(array, value, i=0)
+  indices = array.length - 1
+
+  if i > indices
+    return false
+
+  else
+    if array[i] == value
+      return true
+    else
+      return search(array, value, i += 1)
+    end
+  end
 end
 
-def is_palindrome(s)
-  puts "Not implemented."
+
+#time comp: O(n) where n is the number of elements in the unsorted array; you need to check each one in the worst case.
+#space comp: O(n) where n is the number of elements in the array, as you are calling the function that many times.
+
+
+# Write #7
+# is_palindrome(s) Write a recursive method pal that accepts a string s as a parameter and returns a boolean value indicating if that string is a palindrome or not.
+# e.g. is_palindrome("racecar") = true
+# e.g. is_palindrome("smile") = false Challenge: Try doing this without creating new strings in the process of solving the problem.
+
+def is_palindrome(s, i=0, j=-1)
+  base_case = s.length / 2
+  if s[i] == s[j] || (i == base_case)
```

Comment:
Same bug as with reverse_inplace. "toast" will return true when 't' at i=0 matches the 't' at j=4, but the string is not a palindrome and this method will return true.